lib/s3_inventory.rb

```diff
@@ -0,0 +1,192 @@
+        end
+      ensure
+        connection.exec("DROP TABLE #{table_name}") unless connection.nil?
+      end
+    end
+  end
+
+  def copy_archive_to_tmp_directory
+    log "Downloading archive to tmp directory..."
+    failure_message = "Failed to download archive to tmp directory."
+
+    @s3_helper.download_file(file.key, @archive_filename, failure_message)
+  end
+
+  def unzip_archive
+    log "Unzipping archive, this may take a while..."
+
+    FileUtils.cd(@tmp_directory) do
+      Discourse::Utils.execute_command('gzip', '--decompress', @archive_filename, failure_message: "Failed to unzip archive.")
+    end
+  end
+
+  def update_bucket_policy
+    @s3_helper.s3_client.put_bucket_policy(
+      bucket: bucket_name,
+      policy: {
+        "Version": "2012-10-17",
+        "Statement": [
+          {
+            "Sid": "InventoryAndAnalyticsPolicy",
+            "Effect": "Allow",
+            "Principal": { "Service": "s3.amazonaws.com" },
+            "Action": ["s3:PutObject"],
+            "Resource": ["arn:aws:s3:::#{inventory_path}/*"],
+            "Condition": {
+              "ArnLike": {
+                "aws:SourceArn": "arn:aws:s3:::#{bucket_name}"
+              },
+              "StringEquals": {
+                "s3:x-amz-acl": "bucket-owner-full-control"
+              }
+            }
+          }
+        ]
+      }.to_json
+    )
+  end
+
+  def update_bucket_inventory_configuration
+    @s3_helper.s3_client.put_bucket_inventory_configuration(
+      bucket: bucket_name,
+      id: inventory_id,
+      inventory_configuration: inventory_configuration,
+      use_accelerate_endpoint: false
+    )
+  end
+
+  def inventory_configuration
+    filter_prefix = inventory_id
+    destination_prefix = File.join(INVENTORY_PREFIX, inventory_id)
+
+    if bucket_folder_path.present?
+      filter_prefix = File.join(bucket_folder_path, filter_prefix)
+      destination_prefix = File.join(bucket_folder_path, destination_prefix)
+    end
+
+    {
+      destination: {
+        s3_bucket_destination: {
+          bucket: "arn:aws:s3:::#{bucket_name}",
+          prefix: destination_prefix,
+          format: "CSV"
+        }
+      },
+      filter: {
+        prefix: filter_prefix
+      },
+      is_enabled: SiteSetting.enable_s3_inventory,
+      id: inventory_id,
+      included_object_versions: "Current",
+      optional_fields: ["ETag"],
+      schedule: {
+        frequency: "Daily"
+      }
+    }
+  end
+
+  private
+
+  def bucket_name
+    @s3_helper.s3_bucket_name
+  end
+
+  def bucket_folder_path
+    @s3_helper.s3_bucket_folder_path
+  end
+
+  def unsorted_files
+    objects = []
+
+    @s3_helper.list(File.join(inventory_path, "data")).each do |obj|
+      if obj.key.match?(/\.csv\.gz$/i)
+        objects << obj
+      end
+    end
+
+    objects
+  rescue Aws::Errors::ServiceError => e
+    Rails.logger.warn("Failed to list inventory from S3: #{e.message.presence || e.class.name}")
+    raise StorageError
```

Comment:
Same as above, I don't feel like this custom error helps us in debugging the error in anyway.